modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/disk/ExternalResultHashIndex.java

```diff
@@ -0,0 +1,469 @@
+        try {
+            entriesCnt = 0;
+
+            oldFile.position(0);
+
+            for (long i = 0; i < oldSize; i++) {
+                Entry e = readEntryFromIndexFile(null, oldFile);
+
+                if (!e.isRemoved() && !e.isEmpty())
+                    putEntryToFreeSlot(e.hashCode(), e.rowAddress());
+            }
+        }
+        catch (IOException e) {
+            U.closeQuiet(fileIo);
+            U.closeQuiet(oldFile);
+
+            throw new IgniteException("Failed to extend hash index.", e);
+        }
+        finally {
+            U.closeQuiet(oldFile);
+            oldIdxFile.delete();
+        }
+    }
+
+    /**
+     * Initiates new file.
+     *
+     * @param cap New file capacity in entries (rows). Should be the power of 2.
+     */
+    private void initNewIndexFile(long cap) {
+        try {
+            assert cap > 0 && (cap & (cap - 1)) == 0; // Should be the positive power of 2.
+
+            this.cap = cap;
+
+            idxFile = new File(dir, spillFileName + "_idx_" + id++);
+
+            idxFile.deleteOnExit();
+
+            FileIOFactory fileIOFactory = ctx.query().fileIOFactory();
+
+            fileIo = fileIOFactory.create(idxFile, CREATE_NEW, READ, WRITE);
+
+            // Write empty data to the end of the file to extend it.
+            reusableBuff.clear();
+            fileIo.write(reusableBuff, cap * Entry.ENTRY_BYTES);
+        }
+        catch (IOException e) {
+            throw new IgniteException("Failed to create an index spill file for the intermediate query results.", e);
+        }
+    }
+
+    /**
+     * Hash index entry.
+     */
+    private static class Entry {
+        /** Size of the has index entry in bytes. */
+        static final int ENTRY_BYTES = Integer.BYTES + Long.BYTES;
+
+        /** Hash code. */
+        private final int hashCode;
+
+        /** Address of the row in the rows file. */
+        private final long rowAddr;
+
+        /** Entry slot number. */
+        private final long slot;
+
+        /**
+         * @param hashCode Hash code.
+         * @param rowAddr Address of the row in the rows file.
+         * @param slot Entry slot number.
+         */
+        Entry(int hashCode, long rowAddr, long slot) {
+            this.hashCode = hashCode;
+            this.rowAddr = rowAddr;
+            this.slot = slot;
+        }
+
+        /**
+         * @return Hash code.
+         */
+        public int hashCode() {
+            return hashCode;
+        }
+
+        /**
+         * @return Address of the row in the rows file.
+         */
+        public long rowAddress() {
+            return rowAddr;
+        }
+
+        /**
+         * @return Entry slot number.
+         */
+        public long slot() {
+            return slot;
+        }
+
+        /**
+         * @return Removed flag.
+         */
+        public boolean isRemoved() {
+            return rowAddr == -2;
```

Comment:
Let's introduce a constant for Removed and Empty status.

Reply:
Done